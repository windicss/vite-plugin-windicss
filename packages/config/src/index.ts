import fs from 'node:fs'
import path from 'node:path'
import _debug from 'debug'
import type { JITI } from 'jiti'
import _jiti from 'jiti'
import type { FullConfig as WindiCssOptions } from 'windicss/types/interfaces'

let jiti: JITI

export type { WindiCssOptions }

export const defaultConfigureFiles = [
  'windi.config.ts',
  'windi.config.mts',
  'windi.config.js',
  'windi.config.mjs',
  'windi.config.cjs',

  'windicss.config.ts',
  'windicss.config.mts',
  'windicss.config.js',
  'windicss.config.mjs',
  'windicss.config.cjs',

  'tailwind.config.ts',
  'tailwind.config.mts',
  'tailwind.config.js',
  'tailwind.config.mjs',
  'tailwind.config.cjs',
]

export interface LoadConfigurationOptions {
  /**
   * Name for debug
   *
   * @default 'windicss-config'
   * @internal
   */
  name?: string
  /**
   * Config object or path
   */
  config?: WindiCssOptions | string
  /**
   * CWD
   *
   * @default process.cwd
   * @internal
   */
  root?: string
  /**
   * A list of filename of paths to search of config files
   */
  configFiles?: string[]
  /**
   * On loading configuration error
   *
   * @default [throw error]
   */
  onConfigurationError?: (error: unknown) => void
  /**
   * On configure file not found
   *
   * @default [emit warning]
   */
  onConfigurationNotFound?: (filepath: string) => void
}

export function loadConfiguration(options: LoadConfigurationOptions): {
  config: WindiCssOptions
  filepath?: string
  error?: unknown
} {
  if (!jiti)
    jiti = _jiti(process.cwd(), { requireCache: false, cache: false, v8cache: false })

  let resolved: WindiCssOptions = {}
  let configFilePath: string | undefined
  let error: unknown

  const {
    name = 'windicss-config',
    config,
    root = process.cwd(),
    configFiles: configureFiles = defaultConfigureFiles,
    onConfigurationError = (e) => {
      throw e
    },
    onConfigurationNotFound = (path) => {
      console.warn(`[${name}] config file "${path}" not found, ignored`)
    },
  } = options

  const debugConfig = _debug(`${name}:config`)

  if (typeof config === 'string' || !config) {
    if (!config) {
      for (const name of configureFiles) {
        const tryPath = path.resolve(root, name)
        if (fs.existsSync(tryPath)) {
          configFilePath = tryPath
          break
        }
      }
    }
    else {
      configFilePath = path.resolve(root, config)
      if (!fs.existsSync(configFilePath)) {
        onConfigurationNotFound(config)
        configFilePath = undefined
      }
    }

    if (configFilePath) {
      try {
        debugConfig('loading from ', configFilePath)

        resolved = jiti(configFilePath)
        if (resolved.default)
          resolved = resolved.default
      }
      catch (e) {
        error = e
        configFilePath = undefined
        resolved = {}
        onConfigurationError?.(e)
      }
    }
  }
  else {
    resolved = config
  }

  debugConfig(resolved)

  return {
    error,
    config: resolved,
    filepath: configFilePath,
  }
}
