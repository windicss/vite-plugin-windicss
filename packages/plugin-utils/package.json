{
  "name": "@windicss/plugin-utils",
  "version": "1.8.6",
  "description": "Common utils for building integrations of Windi CSS",
  "author": "antfu <anthonyfu117@hotmail.com>",
  "license": "MIT",
  "funding": "https://github.com/sponsors/antfu",
  "homepage": "https://github.com/antfu/vite-plugin-windicss",
  "repository": {
    "type": "git",
    "url": "https://github.com/antfu/vite-plugin-windicss"
  },
  "bugs": "https://github.com/antfu/vite-plugin-windicss/issues",
  "keywords": [
    "windicss",
    "tailwindcss"
  ],
  "exports": {
    ".": {
      "require": "./dist/index.js",
      "import": "./dist/index.mjs"
    },
    "./*": "./*"
  },
  "main": "dist/index.js",
  "module": "dist/index.mjs",
  "types": "dist/index.d.ts",
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "tsup src/index.ts --dts --format cjs,esm --no-splitting --external sucrase/register/ts,pug,windicss/utils/style,windicss/utils/parser,windicss/utils",
    "dev": "npm run build -- --watch",
    "prepublishOnly": "npm run build"
  },
  "dependencies": {
    "@antfu/utils": "^0.5.2",
    "@windicss/config": "workspace:*",
    "debug": "^4.3.4",
    "fast-glob": "^3.2.11",
    "magic-string": "^0.26.2",
    "micromatch": "^4.0.5",
    "windicss": "^3.5.6"
  },
  "devDependencies": {
    "pug": "^3.0.2"
  }
}
