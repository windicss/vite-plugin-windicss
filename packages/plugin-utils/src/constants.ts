export const defaultAlias: Record<string, TagNames> = {
  'router-link': 'a',
}

export const preflightTags = ['html', 'body', 'div']

export const htmlTags = [
  'html',
  'body',
  'div',
  'a',
  'abbr',
  'address',
  'area',
  'article',
  'aside',
  'audio',
  'base',
  'basefont',
  'bdo',
  'blink',
  'blockquote',
  'br',
  'button',
  'canvas',
  'caption',
  'center',
  'col',
  'colgroup',
  'command',
  'comment',
  'datalist',
  'dd',
  'del',
  'details',
  'dir',
  'dl',
  'dt',
  'embed',
  'fieldset',
  'figure',
  'b',
  'big',
  'i',
  'small',
  'tt',
  'font',
  'footer',
  'form',
  'frame',
  'frameset',
  'head',
  'header',
  'hgroup',
  'h1',
  'h2',
  'h3',
  'h4',
  'h5',
  'h6',
  'hr',
  'isindex',
  'iframe',
  'ilayer',
  'img',
  'input',
  'ins',
  'keygen',
  'keygen',
  'label',
  'layer',
  'legend',
  'li',
  'link',
  'map',
  'mark',
  'marquee',
  'menu',
  'meta',
  'meter',
  'multicol',
  'nav',
  'nobr',
  'noembed',
  'noframes',
  'noscript',
  'object',
  'ol',
  'optgroup',
  'option',
  'output',
  'p',
  'param',
  'cite',
  'code',
  'dfn',
  'em',
  'kbd',
  'samp',
  'strong',
  'var',
  'plaintext',
  'pre',
  'progress',
  'q',
  'ruby',
  'script',
  'section',
  'select',
  'spacer',
  'span',
  's',
  'strike',
  'style',
  'sub',
  'sup',
  'svg',
  'table',
  'tbody',
  'td',
  'textarea',
  'tfoot',
  'th',
  'thead',
  'time',
  'title',
  'tr',
  'u',
  'ul',
  'video',
  'wbr',
  'wbr',
  'xmp',
] as const

export type TagNames = (typeof htmlTags)[number]
