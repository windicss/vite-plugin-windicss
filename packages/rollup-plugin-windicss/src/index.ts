import { Plugin } from 'rollup'
import _debug from 'debug'
import { createUtils, UserOptions } from '@windicss/plugin-utils'
import { createVirtualModuleLoader } from '../../shared/virtual-module'

const NAME = 'rollup-plugin-windicss'

const debug = {
  hmr: _debug(`${NAME}:hmr`),
  css: _debug(`${NAME}:transform:css`),
  group: _debug(`${NAME}:transform:group`),
  alias: _debug(`${NAME}:transform:alias`),
}

function WindiCssRollupPlugin(userOptions: UserOptions = {}): Plugin[] {
  const utils = createUtils(
    userOptions,
    {
      name: NAME,
      onConfigurationError(e) {
        throw e
      },
    },
  )
  const plugins: Plugin[] = []

  plugins.push({
    name: `${NAME}:alias`,
    async transform(code, id) {
      await utils.ensureInit()
      if (!utils.isDetectTarget(id))
        return
      debug.alias(id)
      return utils.transformAlias(code, true)
    },
  })

  if (userOptions.transformGroups !== false) {
    plugins.push({
      name: `${NAME}:groups`,
      async transform(code, id) {
        await utils.ensureInit()
        if (!utils.isDetectTarget(id))
          return
        debug.group(id)
        return utils.transformGroups(code, true)
      },
    })
  }

  plugins.push({
    name: `${NAME}:entry`,

    ...createVirtualModuleLoader({ utils }),
  })

  if (userOptions.transformCSS !== false) {
    plugins.push({
      name: `${NAME}:css`,
      async transform(code, id) {
        await utils.ensureInit()
        if (!utils.isCssTransformTarget(id))
          return
        debug.css(id)
        return utils.transformCSS(code, id)
      },
    })
  }

  return plugins
}

export * from '@windicss/plugin-utils'
export default WindiCssRollupPlugin
