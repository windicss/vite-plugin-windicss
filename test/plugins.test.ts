import { describe, expect, it } from 'vitest'
import { createUtils } from '../packages/plugin-utils/src'

describe('plugins', () => {
  it('aspect-ratio', async () => {
    const utils = createUtils({
      config: {
        plugins: [
          require('windicss/plugin/aspect-ratio'),
        ],
      },
      scan: false,
      preflight: false,
    })
    await utils.init()
    await utils.extractFile('"aspect-none aspect-w-16 aspect-h-9 aspect-16/9"')

    const css = await utils.generateCSS()
    expect(utils.classesGenerated).toMatchSnapshot('classes')
    expect(css).toMatchSnapshot('generated-css')
  })

  it('forms', async () => {
    const utils = createUtils({
      config: {
        plugins: [
          require('windicss/plugin/forms'),
        ],
      },
      preflight: {
        includeBase: false,
      },
      scan: false,
    })
    await utils.init()

    await utils.extractFile('<input type="text"/><input type="number"/><select multiple/>')

    const css = await utils.generateCSS()
    expect(utils.classesGenerated).toMatchSnapshot('classes')
    expect(css).toMatchSnapshot('generated-css')
  })

  it('forms', async () => {
    const utils = createUtils({
      config: {
        plugins: [
          require('windicss/plugin/forms'),
        ],
      },
      preflight: {
        includeBase: false,
        safelist: ['[type="text"]', '[type="number"]'],
      },
      scan: false,
    })
    await utils.init()

    const css = await utils.generateCSS()
    expect(utils.classesGenerated).toMatchSnapshot('classes')
    expect(css).toMatchSnapshot('generated-css')
  })

  // #111
  it('custom classes with parentheses', async () => {
    const utils = createUtils({
      config: {
        theme: {
          extend: {
            height: (theme: any) => ({
              '(custom)': `calc(100% - ${theme('spacing.16')})`,
            }),
          },
        },
      },
      preflight: false,
      scan: false,
    })
    await utils.init()
    await utils.extractFile('class="h-(custom)"')
    const css = await utils.generateCSS()
    expect(utils.classesGenerated.size).toBe(1)
    expect(css).toMatchSnapshot('generated-css')
  })

  it('typography rtl', async () => {
    const utils = createUtils({
      config: {
        plugins: [
          // eslint-disable-next-line @typescript-eslint/no-var-requires
          require('windicss/plugin/typography')({
            rtl: true,
          }),
        ],
      },
      preflight: {
        includeBase: false,
      },
      scan: false,
    })
    await utils.init()
    await utils.extractFile('<div class="prose"/>')

    const css = await utils.generateCSS()
    expect(utils.classesGenerated).toMatchSnapshot('classes')
    expect(css).toMatchSnapshot('generated-css')
  })
})
