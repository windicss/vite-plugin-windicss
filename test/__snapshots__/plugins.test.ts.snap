// Vitest Snapshot v1

exports[`plugins > aspect-ratio > classes 1`] = `
Set {
  "aspect-none",
  "aspect-w-16",
  "aspect-h-9",
  "aspect-16/9",
}
`;

exports[`plugins > aspect-ratio > generated-css 1`] = `
"/* windicss layer base */

/* windicss layer components */
.aspect-none {
  position: static;
  padding-bottom: 0;
}
.aspect-none > * {
  position: static;
  height: auto;
  width: auto;
  top: auto;
  right: auto;
  bottom: auto;
  left: auto;
}
.aspect-w-16 {
  --tw-aspect-w: 16;
  position: relative;
  padding-bottom: calc(var(--tw-aspect-h) / var(--tw-aspect-w) * 100%);
}
.aspect-w-16 > * {
  position: absolute;
  height: 100%;
  width: 100%;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
}
.aspect-h-9 {
  --tw-aspect-h: 9;
}
.aspect-16\\\\/9 {
  position: relative;
  padding-bottom: 177.777778%;
}
.aspect-16\\\\/9 > * {
  position: absolute;
  height: 100%;
  width: 100%;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
}
/* windicss layer utilities */"
`;

exports[`plugins > custom classes with parentheses > generated-css 1`] = `
"/* windicss layer base */

/* windicss layer components */

/* windicss layer utilities */
.h-\\\\(custom\\\\) {
  height: calc(100% - 4rem);
}"
`;

exports[`plugins > forms > classes 1`] = `Set {}`;

exports[`plugins > forms > classes 2`] = `Set {}`;

exports[`plugins > forms > generated-css 1`] = `
"/* windicss layer base */
[type='text'], [type='email'], [type='url'], [type='password'], [type='number'], [type='date'], [type='datetime-local'], [type='month'], [type='search'], [type='tel'], [type='time'], [type='week'], [multiple], textarea, select {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background-color: #fff;
  border-color: #6b7280;
  border-width: 1px;
  border-radius: 0px;
  padding-top: 0.5rem;
  padding-right: 0.75rem;
  padding-bottom: 0.5rem;
  padding-left: 0.75rem;
  font-size: 1rem;
  line-height: 1.5rem;
}
[type='text']:focus, [type='email']:focus, [type='url']:focus, [type='password']:focus, [type='number']:focus, [type='date']:focus, [type='datetime-local']:focus, [type='month']:focus, [type='search']:focus, [type='tel']:focus, [type='time']:focus, [type='week']:focus, [multiple]:focus, textarea:focus, select:focus {
  --tw-ring-inset: var(--tw-empty,/*!*/ /*!*/);
  --tw-ring-offset-width: 0px;
  --tw-ring-offset-color: #fff;
  --tw-ring-color: #2563eb;
  --tw-ring-offset-shadow: var(--tw-ring-inset) 0 0 0 var(--tw-ring-offset-width) var(--tw-ring-offset-color);
  --tw-ring-shadow: var(--tw-ring-inset) 0 0 0 calc(1px + var(--tw-ring-offset-width)) var(--tw-ring-color);
  box-shadow: var(--tw-ring-offset-shadow), var(--tw-ring-shadow), var(--tw-shadow, 0 0 #0000);
  border-color: #2563eb;
}
[type='checkbox'], [type='radio'] {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  padding: 0;
  -webkit-print-color-adjust: exact;
  print-color-adjust: exact;
  display: inline-block;
  vertical-align: middle;
  background-origin: border-box;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  flex-shrink: 0;
  height: 1rem;
  width: 1rem;
  color: #2563eb;
  background-color: #fff;
  border-color: #6b7280;
  border-width: 1px;
}
[type='checkbox']:focus, [type='radio']:focus {
  --tw-ring-inset: var(--tw-empty,/*!*/ /*!*/);
  --tw-ring-offset-width: 2px;
  --tw-ring-offset-color: #fff;
  --tw-ring-color: #2563eb;
  --tw-ring-offset-shadow: var(--tw-ring-inset) 0 0 0 var(--tw-ring-offset-width) var(--tw-ring-offset-color);
  --tw-ring-shadow: var(--tw-ring-inset) 0 0 0 calc(2px + var(--tw-ring-offset-width)) var(--tw-ring-color);
  box-shadow: var(--tw-ring-offset-shadow), var(--tw-ring-shadow), var(--tw-shadow, 0 0 #0000);
  border-color: #6b7280;
}
[type='checkbox']:checked, [type='radio']:checked {
  border-color: transparent;
  background-color: currentColor;
  background-size: 100% 100%;
  background-position: center;
  background-repeat: no-repeat;
}
[type='checkbox']:checked:hover, [type='checkbox']:checked:focus, [type='radio']:checked:hover, [type='radio']:checked:focus {
  border-color: transparent;
  background-color: currentColor;
}
[type='checkbox']:indeterminate:hover, [type='checkbox']:indeterminate:focus {
  border-color: transparent;
  background-color: currentColor;
}
::-webkit-datetime-edit-fields-wrapper {
  padding: 0;
}
::-webkit-date-and-time-value {
  min-height: 1.5em;
}
[multiple] {
  background-image: initial;
  background-position: initial;
  background-repeat: unset;
  background-size: initial;
  padding-right: 0.75rem;
  -webkit-print-color-adjust: unset;
  print-color-adjust: unset;
}
[type='checkbox'] {
  border-radius: 0px;
}
[type='radio'] {
  border-radius: 100%;
}
[type='checkbox']:checked {
  background-image: url(\\"data:image/svg+xml,%3csvg viewBox='0 0 16 16' fill='white' xmlns='http://www.w3.org/2000/svg'%3e%3cpath d='M12.207 4.793a1 1 0 010 1.414l-5 5a1 1 0 01-1.414 0l-2-2a1 1 0 011.414-1.414L6.5 9.086l4.293-4.293a1 1 0 011.414 0z'/%3e%3c/svg%3e\\");
}
[type='radio']:checked {
  background-image: url(\\"data:image/svg+xml,%3csvg viewBox='0 0 16 16' fill='white' xmlns='http://www.w3.org/2000/svg'%3e%3ccircle cx='8' cy='8' r='3'/%3e%3c/svg%3e\\");
}
[type='checkbox']:indeterminate {
  background-image: url(\\"data:image/svg+xml,%3csvg xmlns='http://www.w3.org/2000/svg' fill='none' viewBox='0 0 16 16'%3e%3cpath stroke='white' stroke-linecap='round' stroke-linejoin='round' stroke-width='2' d='M4 8h8'/%3e%3c/svg%3e\\");
  border-color: transparent;
  background-color: currentColor;
  background-size: 100% 100%;
  background-position: center;
  background-repeat: no-repeat;
}
[type='file'] {
  background: unset;
  border-color: inherit;
  border-width: 0;
  border-radius: 0;
  padding: 0;
  font-size: unset;
  line-height: inherit;
}
[type='file']:focus {
  outline: 1px solid ButtonText;
  outline: 1px auto -webkit-focus-ring-color;
}
input::-moz-placeholder, textarea::-moz-placeholder {
  color: #6b7280;
  opacity: 1;
}
input::-ms-input-placeholder, textarea::-ms-input-placeholder {
  color: #6b7280;
  opacity: 1;
}
input::placeholder, textarea::placeholder {
  color: #6b7280;
  opacity: 1;
}
select {
  background-image: url(\\"data:image/svg+xml,%3csvg xmlns='http://www.w3.org/2000/svg' fill='none' viewBox='0 0 20 20'%3e%3cpath stroke='%236b7280' stroke-linecap='round' stroke-linejoin='round' stroke-width='1.5' d='M6 8l4 4 4-4'/%3e%3c/svg%3e\\");
  background-position: right 0.5rem center;
  background-repeat: no-repeat;
  background-size: 1.5em 1.5em;
  padding-right: 2.5rem;
  -webkit-print-color-adjust: exact;
  print-color-adjust: exact;
}
/* windicss layer components */

/* windicss layer utilities */"
`;

exports[`plugins > forms > generated-css 2`] = `
"/* windicss layer base */
[type='text'], [type='email'], [type='url'], [type='password'], [type='number'], [type='date'], [type='datetime-local'], [type='month'], [type='search'], [type='tel'], [type='time'], [type='week'], [multiple], textarea, select {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background-color: #fff;
  border-color: #6b7280;
  border-width: 1px;
  border-radius: 0px;
  padding-top: 0.5rem;
  padding-right: 0.75rem;
  padding-bottom: 0.5rem;
  padding-left: 0.75rem;
  font-size: 1rem;
  line-height: 1.5rem;
}
[type='text']:focus, [type='email']:focus, [type='url']:focus, [type='password']:focus, [type='number']:focus, [type='date']:focus, [type='datetime-local']:focus, [type='month']:focus, [type='search']:focus, [type='tel']:focus, [type='time']:focus, [type='week']:focus, [multiple]:focus, textarea:focus, select:focus {
  --tw-ring-inset: var(--tw-empty,/*!*/ /*!*/);
  --tw-ring-offset-width: 0px;
  --tw-ring-offset-color: #fff;
  --tw-ring-color: #2563eb;
  --tw-ring-offset-shadow: var(--tw-ring-inset) 0 0 0 var(--tw-ring-offset-width) var(--tw-ring-offset-color);
  --tw-ring-shadow: var(--tw-ring-inset) 0 0 0 calc(1px + var(--tw-ring-offset-width)) var(--tw-ring-color);
  box-shadow: var(--tw-ring-offset-shadow), var(--tw-ring-shadow), var(--tw-shadow, 0 0 #0000);
  border-color: #2563eb;
}
[type='checkbox'], [type='radio'] {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  padding: 0;
  -webkit-print-color-adjust: exact;
  print-color-adjust: exact;
  display: inline-block;
  vertical-align: middle;
  background-origin: border-box;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  flex-shrink: 0;
  height: 1rem;
  width: 1rem;
  color: #2563eb;
  background-color: #fff;
  border-color: #6b7280;
  border-width: 1px;
}
[type='checkbox']:focus, [type='radio']:focus {
  --tw-ring-inset: var(--tw-empty,/*!*/ /*!*/);
  --tw-ring-offset-width: 2px;
  --tw-ring-offset-color: #fff;
  --tw-ring-color: #2563eb;
  --tw-ring-offset-shadow: var(--tw-ring-inset) 0 0 0 var(--tw-ring-offset-width) var(--tw-ring-offset-color);
  --tw-ring-shadow: var(--tw-ring-inset) 0 0 0 calc(2px + var(--tw-ring-offset-width)) var(--tw-ring-color);
  box-shadow: var(--tw-ring-offset-shadow), var(--tw-ring-shadow), var(--tw-shadow, 0 0 #0000);
  border-color: #6b7280;
}
[type='checkbox']:checked, [type='radio']:checked {
  border-color: transparent;
  background-color: currentColor;
  background-size: 100% 100%;
  background-position: center;
  background-repeat: no-repeat;
}
[type='checkbox']:checked:hover, [type='checkbox']:checked:focus, [type='radio']:checked:hover, [type='radio']:checked:focus {
  border-color: transparent;
  background-color: currentColor;
}
[type='checkbox']:indeterminate:hover, [type='checkbox']:indeterminate:focus {
  border-color: transparent;
  background-color: currentColor;
}
::-webkit-datetime-edit-fields-wrapper {
  padding: 0;
}
::-webkit-date-and-time-value {
  min-height: 1.5em;
}
[multiple] {
  background-image: initial;
  background-position: initial;
  background-repeat: unset;
  background-size: initial;
  padding-right: 0.75rem;
  -webkit-print-color-adjust: unset;
  print-color-adjust: unset;
}
[type='checkbox'] {
  border-radius: 0px;
}
[type='radio'] {
  border-radius: 100%;
}
[type='checkbox']:checked {
  background-image: url(\\"data:image/svg+xml,%3csvg viewBox='0 0 16 16' fill='white' xmlns='http://www.w3.org/2000/svg'%3e%3cpath d='M12.207 4.793a1 1 0 010 1.414l-5 5a1 1 0 01-1.414 0l-2-2a1 1 0 011.414-1.414L6.5 9.086l4.293-4.293a1 1 0 011.414 0z'/%3e%3c/svg%3e\\");
}
[type='radio']:checked {
  background-image: url(\\"data:image/svg+xml,%3csvg viewBox='0 0 16 16' fill='white' xmlns='http://www.w3.org/2000/svg'%3e%3ccircle cx='8' cy='8' r='3'/%3e%3c/svg%3e\\");
}
[type='checkbox']:indeterminate {
  background-image: url(\\"data:image/svg+xml,%3csvg xmlns='http://www.w3.org/2000/svg' fill='none' viewBox='0 0 16 16'%3e%3cpath stroke='white' stroke-linecap='round' stroke-linejoin='round' stroke-width='2' d='M4 8h8'/%3e%3c/svg%3e\\");
  border-color: transparent;
  background-color: currentColor;
  background-size: 100% 100%;
  background-position: center;
  background-repeat: no-repeat;
}
[type='file'] {
  background: unset;
  border-color: inherit;
  border-width: 0;
  border-radius: 0;
  padding: 0;
  font-size: unset;
  line-height: inherit;
}
[type='file']:focus {
  outline: 1px solid ButtonText;
  outline: 1px auto -webkit-focus-ring-color;
}
input::-moz-placeholder, textarea::-moz-placeholder {
  color: #6b7280;
  opacity: 1;
}
input::-ms-input-placeholder, textarea::-ms-input-placeholder {
  color: #6b7280;
  opacity: 1;
}
input::placeholder, textarea::placeholder {
  color: #6b7280;
  opacity: 1;
}
select {
  background-image: url(\\"data:image/svg+xml,%3csvg xmlns='http://www.w3.org/2000/svg' fill='none' viewBox='0 0 20 20'%3e%3cpath stroke='%236b7280' stroke-linecap='round' stroke-linejoin='round' stroke-width='1.5' d='M6 8l4 4 4-4'/%3e%3c/svg%3e\\");
  background-position: right 0.5rem center;
  background-repeat: no-repeat;
  background-size: 1.5em 1.5em;
  padding-right: 2.5rem;
  -webkit-print-color-adjust: exact;
  print-color-adjust: exact;
}
/* windicss layer components */

/* windicss layer utilities */"
`;

exports[`plugins > typography rtl > classes 1`] = `
Set {
  "prose",
}
`;

exports[`plugins > typography rtl > generated-css 1`] = `
"/* windicss layer base */

/* windicss layer components */
.prose {
  color: #374151;
  max-width: 65ch;
  font-size: 1rem;
  line-height: 1.75;
}
.prose [class~=\\"lead\\"] {
  color: #4b5563;
  font-size: 1.25em;
  line-height: 1.6;
  margin-top: 1.2em;
  margin-bottom: 1.2em;
}
.prose a {
  color: #111827;
  text-decoration: underline;
  font-weight: 500;
}
.prose strong {
  color: #111827;
  font-weight: 600;
}
.prose ol[type=\\"A\\"] {
  --list-counter-style: upper-alpha;
}
.prose ol[type=\\"a\\"] {
  --list-counter-style: lower-alpha;
}
.prose ol[type=\\"A s\\"] {
  --list-counter-style: upper-alpha;
}
.prose ol[type=\\"a s\\"] {
  --list-counter-style: lower-alpha;
}
.prose ol[type=\\"I\\"] {
  --list-counter-style: upper-roman;
}
.prose ol[type=\\"i\\"] {
  --list-counter-style: lower-roman;
}
.prose ol[type=\\"I s\\"] {
  --list-counter-style: upper-roman;
}
.prose ol[type=\\"i s\\"] {
  --list-counter-style: lower-roman;
}
.prose ol[type=\\"1\\"] {
  --list-counter-style: decimal;
}
.prose ol > li {
  position: relative;
  padding-left: 1.75em;
}
.prose ol > li::before {
  content: counter(list-item, var(--list-counter-style, decimal)) \\".\\";
  position: absolute;
  font-weight: 400;
  color: #6b7280;
  left: 0;
}
.prose ul > li {
  position: relative;
  padding-left: 1.75em;
}
.prose ul > li::before {
  content: \\"\\";
  position: absolute;
  background-color: #d1d5db;
  border-radius: 50%;
  width: 0.375em;
  height: 0.375em;
  top: calc(0.875em - 0.1875em);
  left: 0.25em;
}
.prose hr {
  border-color: #e5e7eb;
  margin-top: 3em;
  margin-bottom: 3em;
}
.prose blockquote {
  font-weight: 500;
  font-style: italic;
  color: #111827;
  border-left-width: 0.25rem;
  border-color: #e5e7eb;
  quotes: \\"\\\\201C\\"\\"\\\\201D\\"\\"\\\\2018\\"\\"\\\\2019\\";
  margin-top: 1.6em;
  margin-bottom: 1.6em;
  padding-left: 1em;
}
.prose blockquote p:first-of-type::before {
  content: open-quote;
}
.prose blockquote p:last-of-type::after {
  content: close-quote;
}
.prose h1 {
  color: #111827;
  font-weight: 800;
  font-size: 2.25em;
  margin-top: 0;
  margin-bottom: 0.8888889em;
  line-height: 1.1111111;
}
.prose h2 {
  color: #111827;
  font-weight: 700;
  font-size: 1.5em;
  margin-top: 2em;
  margin-bottom: 1em;
  line-height: 1.3333333;
}
.prose h3 {
  color: #111827;
  font-weight: 600;
  font-size: 1.25em;
  margin-top: 1.6em;
  margin-bottom: 0.6em;
  line-height: 1.6;
}
.prose h4 {
  color: #111827;
  font-weight: 600;
  margin-top: 1.5em;
  margin-bottom: 0.5em;
  line-height: 1.5;
}
.prose figure figcaption {
  color: #6b7280;
  font-size: 0.875em;
  line-height: 1.4285714;
  margin-top: 0.8571429em;
}
.prose code {
  color: #111827;
  font-weight: 600;
  font-size: 0.875em;
}
.prose code::before {
  content: \\"\`\\";
}
.prose code::after {
  content: \\"\`\\";
}
.prose a code {
  color: #111827;
}
.prose pre {
  color: #e5e7eb;
  background-color: #1f2937;
  overflow-x: auto;
  font-size: 0.875em;
  line-height: 1.7142857;
  margin-top: 1.7142857em;
  margin-bottom: 1.7142857em;
  border-radius: 0.375rem;
  padding-top: 0.8571429em;
  padding-right: 1.1428571em;
  padding-bottom: 0.8571429em;
  padding-left: 1.1428571em;
}
.prose pre code {
  background-color: transparent;
  border-width: 0;
  border-radius: 0;
  padding: 0;
  font-weight: 400;
  color: inherit;
  font-size: inherit;
  font-family: inherit;
  line-height: inherit;
}
.prose pre code::before {
  content: none;
}
.prose pre code::after {
  content: none;
}
.prose table {
  width: 100%;
  table-layout: auto;
  text-align: left;
  margin-top: 2em;
  margin-bottom: 2em;
  font-size: 0.875em;
  line-height: 1.7142857;
}
.prose thead {
  color: #111827;
  font-weight: 600;
  border-bottom-width: 1px;
  border-bottom-color: #d1d5db;
}
.prose thead th {
  vertical-align: bottom;
  padding-right: 0.5714286em;
  padding-bottom: 0.5714286em;
  padding-left: 0.5714286em;
}
.prose tbody tr {
  border-bottom-width: 1px;
  border-bottom-color: #e5e7eb;
}
.prose tbody tr:last-child {
  border-bottom-width: 0;
}
.prose tbody td {
  vertical-align: top;
  padding-top: 0.5714286em;
  padding-right: 0.5714286em;
  padding-bottom: 0.5714286em;
  padding-left: 0.5714286em;
}
.prose p {
  margin-top: 1.25em;
  margin-bottom: 1.25em;
}
.prose img {
  margin-top: 2em;
  margin-bottom: 2em;
}
.prose video {
  margin-top: 2em;
  margin-bottom: 2em;
}
.prose figure {
  margin-top: 2em;
  margin-bottom: 2em;
}
.prose figure > * {
  margin-top: 0;
  margin-bottom: 0;
}
.prose h2 code {
  font-size: 0.875em;
}
.prose h3 code {
  font-size: 0.9em;
}
.prose ol {
  margin-top: 1.25em;
  margin-bottom: 1.25em;
  list-style-type: none;
}
.prose ul {
  margin-top: 1.25em;
  margin-bottom: 1.25em;
  list-style-type: none;
}
.prose li {
  margin-top: 0.5em;
  margin-bottom: 0.5em;
}
.prose > ul > li p {
  margin-top: 0.75em;
  margin-bottom: 0.75em;
}
.prose > ul > li > *:first-child {
  margin-top: 1.25em;
}
.prose > ul > li > *:last-child {
  margin-bottom: 1.25em;
}
.prose > ol > li > *:first-child {
  margin-top: 1.25em;
}
.prose > ol > li > *:last-child {
  margin-bottom: 1.25em;
}
.prose ul ul, .prose ul ol, .prose ol ul, .prose ol ol {
  margin-top: 0.75em;
  margin-bottom: 0.75em;
}
.prose hr + * {
  margin-top: 0;
}
.prose h2 + * {
  margin-top: 0;
}
.prose h3 + * {
  margin-top: 0;
}
.prose h4 + * {
  margin-top: 0;
}
.prose thead th:first-child {
  padding-left: 0;
}
.prose thead th:last-child {
  padding-right: 0;
}
.prose tbody td:first-child {
  padding-left: 0;
}
.prose tbody td:last-child {
  padding-right: 0;
}
.prose > :first-child {
  margin-top: 0;
}
.prose > :last-child {
  margin-bottom: 0;
}
.prose[dir=\\"rtl\\"] ol > li {
  padding-left: initial;
  padding-right: 1.75em;
}
.prose[dir=\\"rtl\\"] ol > li::before {
  left: initial;
  right: 0;
}
.prose[dir=\\"rtl\\"] ul > li {
  padding-left: initial;
  padding-right: 1.75em;
}
.prose[dir=\\"rtl\\"] ul > li::before {
  left: initial;
  right: 0.25em;
}
.prose[dir=\\"rtl\\"] blockquote {
  border-left-width: 0;
  border-right-width: 0.25rem;
  padding-left: initial;
  padding-right: 1em;
}
.prose[dir=\\"rtl\\"] table {
  text-align: right;
}
.prose[dir=\\"rtl\\"] thead th:first-child {
  padding-left: initial;
  padding-right: 0;
}
.prose[dir=\\"rtl\\"] thead th:last-child {
  padding-right: initial;
  padding-left: 0;
}
.prose[dir=\\"rtl\\"] tbody td:first-child {
  padding-left: initial;
  padding-right: 0;
}
.prose[dir=\\"rtl\\"] tbody td:last-child {
  padding-right: initial;
  padding-left: 0;
}
/* windicss layer utilities */"
`;
